src/webapp/github.py

```diff
@@ -16,12 +18,70 @@ def mk_github_authstr(user, passwd):
     raw = '%s:%s' % (user, passwd)
     return base64.encodebytes(raw.encode()).decode().replace('\n', '')
 
+
 def api_path2url(api_path, **kw):
     fmtstr = re.sub(r':([a-z]+)\b', r'{\1}', api_path)
     path = fmtstr.format(**kw)
     url = api_baseurl + path
     return url
 
+
+def http_to_dict(success, http_response, throw_error=True):
+    """http json response to dict"""
+
+    if success:
+        return json.load(http_response.fp)
+
+    if throw_error:
+
+        if http_response == "Not Found":
+            raise GithubNotFoundException(f"Request not successful: {str(http_response)}")
+
+        elif http_response == "Reference already exists":
+            raise GithubReferenceExistsException
+
+        else:
+            raise GithubRequestException(f"Request not successful: {str(http_response)}")
+
+    return {}
+
+
+class GithubRequestException(Exception):
+    pass
+
+
+class GithubNotFoundException(GithubRequestException):
+    pass
+
+
+class GithubReferenceExistsException(GithubRequestException):
+    pass
+
+
+class GithubUser:
+
+    def __init__(self, name, email):
+        self.name = name
+        self.email = email
+
+    @classmethod
+    def from_token(cls, token):
+        gh_auth = GitHubAuth(None, token)
+        user_response = http_to_dict(*gh_auth.get_user())
+        email_response = http_to_dict(*gh_auth.get_user_email())
+
+        # Grab the first visible email to use, it doesn't matter
+        visible_email = list(filter(lambda x: x['visibility'] != 'private', email_response))[0]['email']
```

Comment:
is there guaranteed to be at least one visible email?

Reply:
Yes, if a user decides they want their email private then Github will make a hashed public email for the user.